src/test/ui/consts/const-eval/const_fn_ptr_fail.rs

```diff
@@ -0,0 +1,15 @@
+// run-pass
+
+// FIXME: this should not pass
+
+#![feature(const_fn)]
+#![feature(const_fn_ptr)]
+
+fn double(x: usize) -> usize { x * 2 }
+const X: fn(usize) -> usize = double;
+
+const fn bar(x: usize) -> usize {
+    X(x)
```

Comment:
@oli-obk this is the issue that we chatted about. Note that `bar` can be called at run-time, but it will always fail with a constant evaluation error when called in a const context because it tries to call a non-const fn.

I'm not sure how to generate an error in this case.

Reply:
XD welcome to RFC territory. This needs an RFC, because we need surface language changes in order to differentiate between callable and not callable function pointers.

This is what I meant by "we can do it behind a feature gate for experimentation, but we'll never be able to stabilize it.". For all intents and purposes, this is const unsound just like casting pointers to usize or comparing pointers. The pointer ops have been made unsafe for this reason, so maybe we make this unsafe, too? Idk. Seems all fishy without an RFC but as an experimental unstable feature seems fine

Reply:
@oli-obk I've updated the OP with a description of the semantics that this PR implements. I'm not sure why this would be `const` unsound. The `const fn` taking pointers should be parametric over the pointer `const`ness, and we should be able to reject that at type checking time without new syntax AFAICT.

Reply:
Sure, that would be one way to do it, it would mean you can't pass any function pointers, because there are no const function pointers. Not compiling is the current semantics and is totally sound. This is all discussed in my RFC and needs its own RFC. I'm 🤧 right now and my brain is mush. I'd prefer not to write out an explanation right now.

Reply:
Ah I see what you mean now.

Reply:
I am not sure if we need language syntax for this initially, e.g., we could just add a type that users cannot write, add coercions to it from `const fn`s, and then do the const parametricity part. But that's 99% of the work, and at that point not allowing people to write down the type feels like unnecessary. 

Either way, the intent of this PR was never to go through all that trouble.